src/main/java/net/snowflake/ingest/streaming/internal/LiteralQuoteUtils.java

```diff
@@ -11,6 +20,38 @@
  */
 class LiteralQuoteUtils {
 
+  /** Maximum number of unquoted column names to store in cache */
+  static final int UNQUOTED_COLUMN_NAME_CACHE_MAX_SIZE = 30000;
+
+  /** Expiration policy for the cache of unquoted column names */
+  static final Duration UNQUOTED_COLUMN_NAME_CACHE_EXPIRATION = Duration.of(5, ChronoUnit.MINUTES);
+
+  /** Cache storing unquoted column names */
+  private static final LoadingCache<String, String> unquotedColumnNamesCache;
+
+  static {
+    unquotedColumnNamesCache =
+        CacheBuilder.newBuilder()
+            .expireAfterAccess(UNQUOTED_COLUMN_NAME_CACHE_EXPIRATION)
+            .maximumSize(UNQUOTED_COLUMN_NAME_CACHE_MAX_SIZE)
+            .build(
+                new CacheLoader<String, String>() {
+                  @Override
+                  public String load(String key) {
+                    return unquoteColumnNameInternal(key);
+                  }
+                });
+  }
+
+  static String unquoteColumnName(String columnName) {
```

Comment:
add comment?

Reply:
@sfc-gh-tzhang 

Comment added.

I locally micro-benchmarked unquoting of 500M quoted column names yesterday and for quoted columns there is ≈4.5x (≈40sec vs 1380sec) speedup. However, today I found that for simple, not quoted column names, there is ≈50% slowdown (25s vs 38s) - for simple columns it is cheaper to call `unquote` than go through the cache. Since the majority of column names is not quoted, I added `isUnquotingExpensive` heuristics, which just check first and last character (it is very fast) and if they are both quotes, the unquoting is considered expensive and it will cache the result.

Reply:
I ran jmh micro-benchmarks that test unquote performance for various column names. Every iteration cleared the cache and unquoted 10M pre-created column names, here are the results:

```
Column format: COLUMN_1_NAME_EXAMPLE[0-1000]
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  0.315           s/op
JmhBench.benchmarkFromCache        avgt    2  0.377           s/op

Column format: column_1_name_example[0-1000]
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  0.545           s/op
JmhBench.benchmarkFromCache        avgt    2  0.377           s/op

Column format: "column_1_name_example[0-1000]"
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  1.257           s/op
JmhBench.benchmarkFromCache        avgt    2  0.402           s/op

Column format: "column_""1_name""_example[0-1000]"
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  3.401           s/op
JmhBench.benchmarkFromCache        avgt    2  0.387           s/op
```


Depending on the input column name, the improvement is between 0.4x - 10x. Only for the optimal case when no transformation is needed (column is not quoted and user passes upper-cased input), there is a small regression (16%), which will be compensated later when we move to from guava cache to scaffeine. These are ballpark numbers for some example column names, the actual numbers will of course depend on the actual column names, number of quotes and their length, but it gives us some idea. 

In a followup PR, I would like to switch to scaffeine that, according to my preliminary benchmarking, will give us additional performance boost.

Reply:
sounds good, let's make sure the security team is ok with the new library, @sfc-gh-japatel should know the process

Reply:
As long as it has a valid allowed license, i think we are good to go! https://lift.snowflake.com/$viewer.do?sysparm_stack=no&sysparm_sys_id=b1287f0e1b9c691404ad33fe034bcb84

Reply:
Great, thank you @sfc-gh-japatel! If that is the case I will squeeze caffeine cache into 1.1.1.

Reply:
Updated benchmarks for caffeine cache:

```
Column format: COLUMN_1_NAME_EXAMPLE[0-1000]
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  0.321           s/op
JmhBench.benchmarkFromCache        avgt    2  0.115           s/op

Column format: column_1_name_example[0-1000]
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  0.538           s/op
JmhBench.benchmarkFromCache        avgt    2  0.129           s/op


Column format: "column_1_name_example[0-1000]"
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  1.197           s/op
JmhBench.benchmarkFromCache        avgt    2  0.143           s/op

Column format: "column_""1_name""_example[0-1000]"
Benchmark                          Mode  Cnt  Score   Error  Units
JmhBench.benchmarkAlwaysRecompute  avgt    2  3.415           s/op
JmhBench.benchmarkFromCache        avgt    2  0.138           s/op
```

The numbers look much better with caffeine. Now we get ~3x speedup for the optimal case, ~8x speedup for the basic quoted case and 25x speedup for the worst case.